pre_commit/metrics.py

```diff
@@ -0,0 +1,66 @@
+from contextlib import contextmanager
+from dataclasses import dataclass
+from datetime import datetime, timezone
+import shlex
+import subprocess
+from typing import List, Generator, Dict, Optional
+import json
+
+@dataclass
+class MetricRecord:
+    name: str
+    successful: bool
+    start_time: datetime
+    end_time: datetime
+
+
+    def for_json(self) -> Dict[str, object]:
+        return {
+            'name': self.name,
+            'success': self.successful,
+            'startTimeSecs': self.start_time.timestamp(),
+            'endTimeSecs': self.end_time.timestamp(),
+        }
+
+
+class TraceSuccess:
```

Comment:
The usage of this feels slightly awkward, but there wasn't an obvious better option. 🤷

Reply:
Cool -- all those are definitely better. Renamed to `Trace` for now.